includes/Core/Util/WP_Context_Switcher.php

```diff
@@ -0,0 +1,124 @@
+<?php
+/**
+ * Class Google\Site_Kit\Core\Util\WP_Context_Switcher
+ *
+ * @package   Google\Site_Kit\Core\Util
+ * @copyright 2020 Google LLC
+ * @license   https://www.apache.org/licenses/LICENSE-2.0 Apache License 2.0
+ * @link      https://sitekit.withgoogle.com
+ */
+
+namespace Google\Site_Kit\Core\Util;
+
+/**
+ * Class for temporarily switching WordPress context, e.g. from admin to frontend.
+ *
+ * @since n.e.x.t
+ * @access private
+ * @ignore
+ */
+class WP_Context_Switcher {
+
+	/**
+	 * Stack of original contexts switched from.
+	 *
+	 * @since n.e.x.t
+	 * @var array
+	 */
+	private static $context_stack = array();
+
+	/**
+	 * Switches to WordPress frontend context if necessary.
+	 *
+	 * Context is only switched if WordPress is not already in frontend context. Context should only ever be switched
+	 * temporarily. Call the returned closure as soon as possible after to restore the original context.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return callable Closure that restores context and returns true if context was restored or false otherwise.
+	 */
+	public static function with_frontend_context() {
+		global $current_screen;
+
+		if ( ! is_admin() ) {
+			return function() {
+				return false;
+			};
+		}
+
+		self::switch_current_screen( 'front' );
+		return self::get_restore_closure();
+	}
+
+	/**
+	 * Switches to WordPress admin context if necessary.
+	 *
+	 * Context is only switched if WordPress is not already in admin context. Context should only ever be switched
+	 * temporarily. Call the returned closure as soon as possible after to restore the original context.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return callable Closure that restores context and returns true if context was restored or false otherwise.
+	 */
+	public static function with_admin_context() {
+		global $current_screen;
+
+		if ( is_admin() ) {
+			return function() {
+				return false;
+			};
+		}
+
+		self::switch_current_screen( 'index' );
+		return self::get_restore_closure();
+	}
+
+	/**
+	 * Switches the current WordPress screen via the given screen ID or hook name.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $screen_id WordPress screen ID.
+	 */
+	private static function switch_current_screen( $screen_id ) {
+		global $current_screen;
+
+		self::load_wp_screen_api();
+		self::$context_stack[] = $current_screen;
+		$current_screen        = \WP_Screen::get( $screen_id ); // phpcs:ignore WordPress.WP.GlobalVariablesOverride.Prohibited
+	}
+
+	/**
+	 * Gets closure to restore previous WordPress context if it has been switched.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return callable Closure that restores context and returns true if context was restored or false otherwise.
```

Comment:
I don't think the return value of the restore function is necessary – at least as a boolean. It should always restore the screen (it's simply setting the global which has zero cost), so the return value is not really useful as a success/failure if it always works 🙂 . If anything, it could return the restored screen instance, so touching `$current_screen` after that point wouldn't be necessary.

Reply:
I think it makes sense to return this for knowing whether context was switched at all. `true|false` here doesn't indicate any sort of error - the method always behaves like it should, but the return value indicates whether a switch was even performed. Which is certainly more relevant for tests than production code, but I think it's worth having.